src/buckets.rs

```diff
@@ -0,0 +1,1277 @@
+//! The low-level primitive behind [`Vec`](crate::Vec): a lazily-initialized array implemented as a
+//! sequence of buckets with sizes of increasing powers of two.
+
+use ::alloc::alloc::{self, alloc_zeroed, handle_alloc_error};
+use ::alloc::boxed::Box;
+use ::alloc::vec::{self, Vec};
+use core::fmt::{self, Debug, Formatter};
+use core::hint::assert_unchecked;
+use core::iter::FusedIterator;
+use core::num::NonZero;
+use core::ptr;
+use core::slice;
+use core::sync::atomic;
+
+use crate::loom::atomic::AtomicPtr;
+use crate::loom::AtomicMut as _;
+
+/// The low-level primitive behind [`Vec`](crate::Vec): a lazily-initialized array implemented as a
+/// sequence of buckets with sizes of increasing powers of two.
+///
+/// The `BUCKETS` generic parameter controls the maximum capacity, and the inline size, of the
+/// type. See [`buckets_for_index_bits`] for a convenient way to calculate its desired value.
+///
+/// # Notes
+///
+/// This type is potentially quite large, so it is advised to put it behind a `Box` or `Arc` or
+/// similar.
```

Comment:
I think this comment can be removed given that it's clear from the `BUCKETS` parameter (but the comment is still relevant on `Vec`).